language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```
@@ -77,58 +66,172 @@ import           Data.HashMap.Monoidal
+   along the arity reconstructing the term.  When it can't find an argument of
+   the appropriate kind it looks for one of the other kind (which should be the
+   one supplied by the user): if it finds one it adds an extra application or
+   instantiation as appropriate to what it's constructed so far and returns the
+   result.  If there are no arguments of either kind left it just returns what
+   it has at that point.  The only circumstances where this is currently called
+   is if (a) the machine is returning a partially applied builtin, or (b) a
+   wrongly interleaved builtin application is being reported in an error.  Note
+   that we don't call this function if a builtin fails for some reason like
+   division by zero; the term is discarded in that case anyway (see [Note:
+   Ignoring context in UserEvaluationError] in Exception.hs)
+-}
+mkBuiltinApplication :: ExMemory -> BuiltinName -> Arity -> [TypeWithMem uni] -> [TermWithMem uni] -> TermWithMem uni
+mkBuiltinApplication ex bn arity0 tys0 args0 =
+  go arity0 tys0 args0 (Builtin ex bn)
+    where go arity tys args term =
+              case (arity, args, tys) of
+                ([], [], [])                        -> term   -- We've got to the end and successfully constructed the entire application
+                (TermArg:arity', arg:args', _)      -> go arity' tys args' (Apply ex term arg)  -- got an expected term argument
+                (TermArg:_,      [],       ty:_)    -> TyInst ex term ty                        -- term expected, type found
+                (TypeArg:arity', _,        ty:tys') -> go arity' tys' args (TyInst ex term ty)  -- got an expected type argument
+                (TypeArg:_,      arg:_,    [])      -> Apply ex term arg                        -- type expected, term found
+                _                                   -> term                                     -- something else, including partial application
```

Maybe throw an error when a `TermArg` is expected, but none is available? And same for `TypeArg`.

---

Wait, I misunderstood the whole thing, I think. I'll study it more.

---

This is all just a bit of a mish-mash.  It's trying to handle the normal case and the error cases at once, so it just does as much as it can and then returns whatever it's managed to come up with.  I think it does the correct thing in the cases we're likely to encounter though.